### DISCUSSION TITLE
rust integration test, how to get AppHandle

### DISCUSSION THREAD

**Discussion by dassio** [tauri-apps/tauri#7456] [Q&A] [answered]
I want to test the Commands implementation in `lib.rs`, in normal running time, I can get the the `AppHandle` by Command injection  
How can I get the `Apphandle` in the test crate under `src-tauri\tests\integration_test.rs`?

**Comment by FabianLars** [comment, marked as answer]
I've never really looked into the test module myself yet but i think the idea is to either use https://docs.rs/tauri/latest/tauri/test/fn.assert_ipc_response.html or https://docs.rs/tauri/latest/tauri/test/fn.mock_app.html (plus `app.handle()`)

**Reply by dassio** [reply]
thanks this seems to be working with normal commands function 
![event_debug](https://github.com/tauri-apps/tauri/assets/733502/e608e7ca-fa5e-43e6-ae25-4a258669790b)

but the event handler does not seem to work, the example code is here: https://github.com/dassio/learning_tauri
```
pub fn create_app<R: tauri::Runtime>(builder: tauri::Builder<R>) -> tauri::App<R> {
    builder
    .setup(|app| {
        app.listen_global("new_event",  move |_event|{
            println!("event handler called")
        });
        Ok(())
    })
        .invoke_handler(tauri::generate_handler![greet])
        .build(tauri::generate_context!())
        .expect("error while running tauri application")
}

#[tauri::command]
fn greet<R: tauri::Runtime>(name: &str, app_handle: AppHandle<R>) -> String {
    app_handle.emit_all("new_event", {}).unwrap();
    
    if name.len() == 0 || name.trim().len() == 0{
        return "please input your name".to_string();
    }

   format!("Hello, {}!", name)
}
```

 `handle_picture_imported_event` never  got called with the test code:
```
#[test]
fn greet_empty_name_correct() {
    let app = create_app(mock_builder());
    let window = app.get_window("main").unwrap();

    assert_ipc_response(
        &window,
        tauri::InvokePayload {
            cmd: "greet".to_string(),
            tauri_module: None,
            callback: tauri::api::ipc::CallbackFn(0),
            error: tauri::api::ipc::CallbackFn(1),
            inner: json!({
                "name" : ""
            }),
        },
        Ok("please input your name"),
    );
}
```

**Comment by dassio** [comment]
as answered in discord by Lucs , I should use `trigger_global` in `app_handle.trigger_global("new_event", Some("test".to_string()));` 
as suggested in : https://docs.rs/tauri/1.4.1/tauri/trait.Manager.html#method.emit_all
